### PR TITLE
delete secrets created by nctl when deleting applications

### PR DESCRIPTION
**Pull request by thirdeyenick**
This will delete git auth secrets which were created by `nctl` on application delete. If the secret is still referenced by another application, a corresponding message will be printed.

### CODE REVIEW

File: api/util/apps.go [hunk 1]
```diff
@@ -11,6 +11,8 @@ import (
 
 const (
 	ApplicationNameLabel = "application.apps.nine.ch/name"
+	ManagedByAnnotation  = "app.kubernetes.io/managed-by"
+	NctlName             = "nctl"
```
Comment (ctrox) [inline api/util/apps.go:15]: We should probably have this const somewhere we can use it from all packages? Grepping the project I already found two existing instances of the command name `"nctl"`:

```
main.go
54:             kong.Name("nctl"),

auth/login.go
28:     NctlExtensionName = "nctl"
```

Not yet sure where it fits best though.

File: delete/application.go [hunk 2]
```diff
@@ -27,12 +32,117 @@ func (app *applicationCmd) Run(ctx context.Context, client *api.Client) error {
...
+			return nil
+		}
+	}
+	if err := client.Delete(ctx, &secret); err != nil {
+		return checkManuallyError(
+			fmt.Errorf(
+				"error when deleting git auth secret %q: %w",
+				secret.Name,
+				err,
+			),
```
Comment (ctrox) [inline delete/application.go:143]: Nit: This could probably fit on one line just fine?

File: delete/application_test.go [hunk 3]
```diff
@@ -5,41 +5,198 @@ import (
...
+					appOne,
+					nctlSecret,
+					noDeletionExpected(customSecret),
+				)
+			}(),
+			name:          "dev",
+			errorExpected: false,
+		},
+		// a secret which has the same name as the application, but no
+		// nctl annotation will no be deleted
```
Comment (ctrox) [inline delete/application_test.go:94]:
```suggestion
		// nctl annotation will not be deleted
```

File: delete/application_test.go [hunk 4]
```diff
@@ -5,41 +5,198 @@ import (
...
+					appOne,
+					noDeletionExpected(appTwo),
+					noDeletionExpected(secret),
+				)
+			}(),
+			name:          "dev",
+			errorExpected: false,
+		},
+		// here we have 2 secrets. One secret created by nctl which is
+		// not in use and another secret non nctl managed secret which
```
Comment (ctrox) [inline delete/application_test.go:69]:
```suggestion
		// not in use and another non nctl managed secret which
```

File: api/util/apps.go (hunk 1 above)
Comment (thirdeyenick) [reply api/util/apps.go:15]: I put it into the `util` package, but I am open for any other location

File: delete/application.go [hunk 5]
```diff
@@ -27,12 +32,117 @@ func (app *applicationCmd) Run(ctx context.Context, client *api.Client) error {
...
+			return nil
+		}
+		return checkManuallyError(fmt.Errorf("error when checking git auth secret %q for application", secret.Name))
+	}
+	managedBy, exists := secret.Annotations[util.ManagedByAnnotation]
+	if !exists || managedBy != util.NctlName {
+		fmt.Printf("will not delete git auth secret %q as it does not contain expected annotation %q",
+			secret.Name,
+			fmt.Sprintf("%s: %s", util.ManagedByAnnotation, util.NctlName),
+		)
```
[the remark this answers is not in the document]
Comment (thirdeyenick) [reply delete/application.go:112]: I can also leave it out. The 2 mentioned cases ("secret with git auth creds made by user" and "secret with other use case made by user") are hard to differentiate. I have no strong preference here.

[the remark this answers is not in the document]
Comment (thirdeyenick) [reply delete/application.go]: I removed it for now.

**Review by ctrox — APPROVED** [review]

**Pull request merged**